### PR TITLE
Add PredicateAdapter and TriggerClauseAdapter

### PR DESCRIPTION
**Pull request by yongpingchen**
- Add the following 2 adapters for serialize/deserialize Predicate/TriggerClause using Gson. 
   - PredicateAdapter
   - TriggerClauseAdapter
- Add small tests for the adapters

### CODE REVIEW

File: thingif/src/main/java/com/kii/thingif/internal/gson/PredicateAdapter.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        if (src instanceof StatePredicate) {
+            JsonElement clause = gson.toJsonTree(
+                    ((StatePredicate) src).getCondition().getClause(),
+                    TriggerClause.class);
+            JsonObject json = new JsonObject();
+            json.add("condition", clause);
+            json.addProperty("triggersWhen", ((StatePredicate) src).getTriggersWhen().name());
+            json.addProperty("eventSource", src.getEventSource().name());
+            return json;
+        }else {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/internal/gson/PredicateAdapter.java]: This is trivial cosmetic issue but could you add space front of `else`? :)

File: thingif/src/main/java/com/kii/thingif/internal/gson/PredicateAdapter.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            .create();
+    @Override
+    public JsonElement serialize(Predicate src, Type typeOfSrc, JsonSerializationContext context) {
+        if (src == null) return null;
+
+        if (src instanceof StatePredicate) {
+            JsonElement clause = gson.toJsonTree(
+                    ((StatePredicate) src).getCondition().getClause(),
+                    TriggerClause.class);
+            JsonObject json = new JsonObject();
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/internal/gson/PredicateAdapter.java]: This is trivial cosmetic issue but could you rename `json` to `ret`? In `else` clause, return value name is `ret`.

File: thingif/src/main/java/com/kii/thingif/internal/gson/TriggerClauseAdapter.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        JsonDeserializer<TriggerClause>{
+
+    private JSONObject triggerClauseToJson(TriggerClause clause) throws JSONException {
+        JSONObject ret = new JSONObject();
+        if (clause instanceof EqualsClauseInTrigger) {
+            EqualsClauseInTrigger eq = (EqualsClauseInTrigger)clause;
+            ret.put("type", "eq");
+            ret.put("alias", eq.getAlias());
+            ret.put("field", eq.getField());
+            ret.put("value", eq.getValue());
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/internal/gson/TriggerClauseAdapter.java]: Why couldn't we use `gson.toJsonTree(src).getAsJsonObject()` here?

**Comment by shinsuke-nara** [comment]
@yongpingchen Could you check my comments?

File: thingif/src/main/java/com/kii/thingif/internal/gson/TriggerClauseAdapter.java (hunk 3 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/internal/gson/TriggerClauseAdapter.java]: @shinsuke-nara because, in `src` object, there is not `type` info.

Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/internal/gson/TriggerClauseAdapter.java]: @yongpingchen 

Like [here](https://github.com/KiiPlatform/thing-if-AndroidSDK/pull/231/files#diff-933a843812e0fc61d70245821974fe64R42), How about following code?

```java
JsonObject ret = gson.toJsonTree(src).getAsJsonObject();
ret.addProperty("type", "eq");
```

Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/internal/gson/TriggerClauseAdapter.java]: fixed on commit(40b4654)

**Comment by yongpingchen** [comment]
@shinsuke-nara  fixed by your comments. Please confirm.

**Comment by shinsuke-nara** [comment]
@yongping LGTM!

**Pull request merged**